File: PathoFroh/src/main/java/com/patho/main/action/dialog/task/ArchiveTaskDialog.java
package com.patho.main.action.dialog.task;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.annotation.Configurable;

import com.patho.main.action.dialog.AbstractDialog;
import com.patho.main.action.handler.GlobalEditViewHandler;
import com.patho.main.action.handler.WorklistViewHandlerAction;
import com.patho.main.common.Dialog;
import com.patho.main.model.patient.Task;
import com.patho.main.service.TaskService;

import lombok.AccessLevel;
import lombok.Getter;
import lombok.Setter;

@Configurable
@Getter
@Setter
public class ArchiveTaskDialog extends AbstractDialog {

	@Autowired
	@Getter(AccessLevel.NONE)
	@Setter(AccessLevel.NONE)
	private WorklistViewHandlerAction worklistViewHandlerAction;

	@Autowired
	@Getter(AccessLevel.NONE)
	@Setter(AccessLevel.NONE)
	private GlobalEditViewHandler globalEditViewHandler;

	@Autowired
	@Getter(AccessLevel.NONE)
	@Setter(AccessLevel.NONE)
	private TaskService taskService;

	/**
	 * If true the task will be removed from worklist
	 */
	private boolean removeFromWorklist;

	/**
	 * commentary for restoring
	 */
	private String commentary;

	/**
	 * True if task was archived
	 */
	private boolean archiveSuccessful;

	public void initAndPrepareBean(Task task) {
		if (initBean(task))
			prepareDialog();
	}

	public boolean initBean(Task task) {
		this.removeFromWorklist = true;

		super.initBean(task, Dialog.TASK_ARCHIVE);

		return true;
	}

	public void archiveTask() {
		try {

			taskService.archiveTask(getTask());

			if (removeFromWorklist) {
				// only remove from worklist if patient has one active task
				if (task.getPatient().getTasks().stream().filter(p -> !p.isFinalized()).count() > 1) {
					mainHandlerAction.sendGrowlMessagesAsResource("growl.error",
							"growl.error.worklist.remove.moreActive");
				} else {
					worklistViewHandlerAction.removePatientFromCurrentWorklist(task.getPatient());
					worklistViewHandlerAction.onDeselectPatient(true);
				}
			}

			mainHandlerAction.sendGrowlMessagesAsResource("growl.task.archived", "growl.task.archived.text");
			setArchiveSuccessful(true);
		} catch (Exception e) {
			onDatabaseVersionConflict();
		}
	}

	public void hideDialog() {
		super.hideDialog(new Boolean(archiveSuccessful));
	}
}

File: PathoFroh/src/main/kotlin/com/patho/main/dialog/task/ArchiveTaskDialog.kt
package com.patho.main.dialog.task

import com.patho.main.action.handler.MessageHandler
import com.patho.main.common.Dialog
import com.patho.main.common.PredefinedFavouriteList
import com.patho.main.dialog.AbstractTaskDialog
import com.patho.main.model.ListItem
import com.patho.main.model.patient.Task
import com.patho.main.repository.jpa.ListItemRepository
import com.patho.main.repository.jpa.TaskRepository
import com.patho.main.service.TaskService
import com.patho.main.ui.transformer.DefaultTransformer
import com.patho.main.util.dialog.event.PatientReloadEvent
import com.patho.main.util.dialog.event.RemovePatientFromWorklistEvent
import com.patho.main.util.dialog.event.TaskReloadEvent
import com.patho.main.util.status.ExtendedNotificationStatus
import com.patho.main.util.status.TotalTaskStatus
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.context.annotation.Scope
import org.springframework.stereotype.Component

@Component()
@Scope(value = "session")
open class ArchiveTaskDialog @Autowired constructor(
        private val taskService: TaskService,
        private val listItemRepository: ListItemRepository,
        private val taskRepository: TaskRepository) : AbstractTaskDialog(Dialog.TASK_ARCHIVE) {

    /**
     * Status object of the task
     */
    open lateinit var taskStatus: TotalTaskStatus

    /**
     * Selected status for displaying infos
     */
    open var selectedReportIntentStatus: ExtendedNotificationStatus.ReportNotificationIntentStatus.ReportIntentStatus? = null

    /**
     * Contains a list of default reasons for filing a task
     */
    open lateinit var predefinedListItems: List<ListItem>

    /**
     * Transformer for list items, used by selectOneMenu
     */
    open lateinit var predefinedListItemTransformer: DefaultTransformer<ListItem>

    /**
     * Selected list item
     */
    open var selectedListItem: ListItem? = null

    /**
     * If true the archivation button will be rendered nevertheless archivation block favourite lists are present
     */
    open var forceFiling = false

    /**
     * Commentary for task archivation
     */
    open var commentary: String = ""

    /**
     * True if the patient containing the task should be removed from the current worklist
     */
    open var removeFromWorklist: Boolean = true

    /**
     * True if filing is possible, either task is completed or it is forced
     */
    open val isFilingPossible
        get() = forceFiling || taskStatus.isArchiveAble

    override fun initBean(task: Task): Boolean {
        val tmp = taskRepository.findByID(task, true, true, true, true, true)

        taskStatus = TotalTaskStatus(tmp)

        predefinedListItems = listItemRepository.findByListTypeAndArchivedOrderByIndexInListAsc(ListItem.StaticList.TASK_ARCHIVE, false)

        predefinedListItemTransformer = DefaultTransformer(predefinedListItems)

        selectedReportIntentStatus = null

        forceFiling = false

        commentary = ""

        removeFromWorklist = !(task.patient?.tasks?.any { it.favouriteLists.any { it.id == PredefinedFavouriteList.NotificationList.id } }
                ?: true)

        return super.initBean(tmp)
    }

    /**
     * Called if a preset from the ListItems is selected. Copies the value auf the list Item into the
     * comment field
     */
    fun onSelectListItem() {
        if (selectedListItem != null)
            commentary = selectedListItem?.value ?: ""
    }

    /**
     * Archives the task and hides the dialog
     */
    fun archiveAndHide() {
        taskService.archiveTask(task, commentary)

        if (removeFromWorklist)
            super.hideDialog(RemovePatientFromWorklistEvent(task.patient!!))
        else
            super.hideDialog(PatientReloadEvent(task.patient!!))

        MessageHandler.sendGrowlMessagesAsResource("growl.task.archived.headline", "growl.task.archived.text")
    }
}